### ISSUE TITLE
Suggestion: fallback weight calculation and armor configs/datapacks

### ISSUE THREAD

**Issue by No3371** [AnOpenSauceDev/Weight#2]
Hi there! Cool mod, I just want to share my two cents as a modpack dev.

Armor weight mods are often designed with a simple "more armor points means more weight" system, which has its pros and cons, like it's simple to implement but it eliminates the possibility of expensive but tough and light armor like dragon leather armor or something.

There are a lot of modpack devs / server devs dedicated to making mods are compatible by creating configuration or datapacks. It'd be very helpful for users if the mod allows configuring a list consists of entries like `magistuarmory:armet=1`. Perhaps you could even encourage people to share their configuratios for mods on Github's Discussion forum.

In summary, I believe the mod would be perfect if it had a fallback weight calculation (1 armor = ? weight) that automatically added weight to all armor items, as well as customizable weight configs instead of allowing only modders to set their items' weight.

**Comment by AnOpenSauceDev** [comment]
This is a pretty sound idea, and i was planning on doing some work on it later this week anyways. I'll let you know if I get it done.

**Comment by AnOpenSauceDev** [comment]
@No3371 

I finally got around to it this afternoon, and now it's done! I'll update the documentation soon, but the example one in the source code will be pretty easy to understand.

**Comment by No3371** [comment]
Neat! That's pretty fast. I'll give it a try after some mods blocking my pack from launching resolved their compatibility issue.

**Comment by No3371** [comment]
Hi, my server failed t launch with the new version, unfortunately with my limited knowledge about MC modding I can't even guess what's wrong:

MC 1.19.2
Fabric API 0.14.19

```
[19:31:19] [main/WARN]: Failed to load datapacks, can't proceed with server load. You can either fix your datapacks or reset to vanilla with --safeMode
java.util.concurrent.ExecutionException: java.lang.NoClassDefFoundError: net/minecraft/class_7923
        at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396) ~[?:?]
        at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073) ~[?:?]
        at net.minecraft.server.Main.main(Main.java:194) [server-intermediary.jar:?]
        at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:462) [fabric-loader-0.14.19.jar:?]
        at net.fabricmc.loader.impl.launch.knot.Knot.launch(Knot.java:74) [fabric-loader-0.14.19.jar:?]
        at net.fabricmc.loader.impl.launch.knot.KnotServer.main(KnotServer.java:23) [fabric-loader-0.14.19.jar:?]
        at net.fabricmc.loader.impl.launch.server.FabricServerLauncher.main(FabricServerLauncher.java:69) [fabric-loader-0.14.19.jar:?]
        at net.fabricmc.installer.ServerLauncher.main(ServerLauncher.java:69) [fabric-server-mc.1.19.2-loader.0.14.19-launcher.0.11.2.jar:0.11.2]
Caused by: java.lang.NoClassDefFoundError: net/minecraft/class_7923
        at an.opensauce.armourweight.api.ServerResourceListener.reload(ServerResourceListener.java:70) ~[Armour%20Weight%201.4.jar:?]
        at net.minecraft.resource.SynchronousResourceReloader.method_29490(SynchronousResourceReloader.java:15) ~[server-intermediary.jar:?]
        at java.util.concurrent.CompletableFuture$UniRun.tryFire(CompletableFuture.java:787) ~[?:?]
        at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:482) ~[?:?]
        at net.minecraft.resource.SimpleResourceReload.method_18365(SimpleResourceReload.java:69) ~[server-intermediary.jar:?]
        at net.minecraft.util.Util.waitAndApply(Util.java:958) ~[server-intermediary.jar:?]
        at net.minecraft.util.Util.waitAndApply(Util.java:946) ~[server-intermediary.jar:?]
        at net.minecraft.server.Main.main(Main.java:156) ~[server-intermediary.jar:?]
        ... 5 more
Caused by: java.lang.ClassNotFoundException: net.minecraft.class_7923
        at jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641) ~[?:?]
        at java.lang.ClassLoader.loadClass(ClassLoader.java:520) ~[?:?]
        at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.loadClass(KnotClassDelegate.java:226) ~[fabric-loader-0.14.19.jar:?]
        at net.fabricmc.loader.impl.launch.knot.KnotClassLoader.loadClass(KnotClassLoader.java:112) ~[fabric-loader-0.14.19.jar:?]
        at java.lang.ClassLoader.loadClass(ClassLoader.java:520) ~[?:?]
        at an.opensauce.armourweight.api.ServerResourceListener.reload(ServerResourceListener.java:70) ~[Armour%20Weight%201.4.jar:?]
        at net.minecraft.resource.SynchronousResourceReloader.method_29490(SynchronousResourceReloader.java:15) ~[server-intermediary.jar:?]
        at java.util.concurrent.CompletableFuture$UniRun.tryFire(CompletableFuture.java:787) ~[?:?]
        at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:482) ~[?:?]
        at net.minecraft.resource.SimpleResourceReload.method_18365(SimpleResourceReload.java:69) ~[server-intermediary.jar:?]
        at net.minecraft.util.Util.waitAndApply(Util.java:958) ~[server-intermediary.jar:?]
        at net.minecraft.util.Util.waitAndApply(Util.java:946) ~[server-intermediary.jar:?]
        at net.minecraft.server.Main.main(Main.java:156) ~[server-intermediary.jar:?]
        ... 5 more
```

**Comment by AnOpenSauceDev** [comment]
> Hi, my server failed t launch with the new version, unfortunately with my limited knowledge about MC modding I can't even guess what's wrong:
> 
> MC 1.19.2 Fabric API 0.14.19
> 
> ```
> [19:31:19] [main/WARN]: Failed to load datapacks, can't proceed with server load. You can either fix your datapacks or reset to vanilla with --safeMode
> java.util.concurrent.ExecutionException: java.lang.NoClassDefFoundError: net/minecraft/class_7923
>         at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396) ~[?:?]
>         at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073) ~[?:?]
>         at net.minecraft.server.Main.main(Main.java:194) [server-intermediary.jar:?]
>         at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:462) [fabric-loader-0.14.19.jar:?]
>         at net.fabricmc.loader.impl.launch.knot.Knot.launch(Knot.java:74) [fabric-loader-0.14.19.jar:?]
>         at net.fabricmc.loader.impl.launch.knot.KnotServer.main(KnotServer.java:23) [fabric-loader-0.14.19.jar:?]
>         at net.fabricmc.loader.impl.launch.server.FabricServerLauncher.main(FabricServerLauncher.java:69) [fabric-loader-0.14.19.jar:?]
>         at net.fabricmc.installer.ServerLauncher.main(ServerLauncher.java:69) [fabric-server-mc.1.19.2-loader.0.14.19-launcher.0.11.2.jar:0.11.2]
> Caused by: java.lang.NoClassDefFoundError: net/minecraft/class_7923
>         at an.opensauce.armourweight.api.ServerResourceListener.reload(ServerResourceListener.java:70) ~[Armour%20Weight%201.4.jar:?]
>         at net.minecraft.resource.SynchronousResourceReloader.method_29490(SynchronousResourceReloader.java:15) ~[server-intermediary.jar:?]
>         at java.util.concurrent.CompletableFuture$UniRun.tryFire(CompletableFuture.java:787) ~[?:?]
>         at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:482) ~[?:?]
>         at net.minecraft.resource.SimpleResourceReload.method_18365(SimpleResourceReload.java:69) ~[server-intermediary.jar:?]
>         at net.minecraft.util.Util.waitAndApply(Util.java:958) ~[server-intermediary.jar:?]
>         at net.minecraft.util.Util.waitAndApply(Util.java:946) ~[server-intermediary.jar:?]
>         at net.minecraft.server.Main.main(Main.java:156) ~[server-intermediary.jar:?]
>         ... 5 more
> Caused by: java.lang.ClassNotFoundException: net.minecraft.class_7923
>         at jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641) ~[?:?]
>         at java.lang.ClassLoader.loadClass(ClassLoader.java:520) ~[?:?]
>         at net.fabricmc.loader.impl.launch.knot.KnotClassDelegate.loadClass(KnotClassDelegate.java:226) ~[fabric-loader-0.14.19.jar:?]
>         at net.fabricmc.loader.impl.launch.knot.KnotClassLoader.loadClass(KnotClassLoader.java:112) ~[fabric-loader-0.14.19.jar:?]
>         at java.lang.ClassLoader.loadClass(ClassLoader.java:520) ~[?:?]
>         at an.opensauce.armourweight.api.ServerResourceListener.reload(ServerResourceListener.java:70) ~[Armour%20Weight%201.4.jar:?]
>         at net.minecraft.resource.SynchronousResourceReloader.method_29490(SynchronousResourceReloader.java:15) ~[server-intermediary.jar:?]
>         at java.util.concurrent.CompletableFuture$UniRun.tryFire(CompletableFuture.java:787) ~[?:?]
>         at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:482) ~[?:?]
>         at net.minecraft.resource.SimpleResourceReload.method_18365(SimpleResourceReload.java:69) ~[server-intermediary.jar:?]
>         at net.minecraft.util.Util.waitAndApply(Util.java:958) ~[server-intermediary.jar:?]
>         at net.minecraft.util.Util.waitAndApply(Util.java:946) ~[server-intermediary.jar:?]
>         at net.minecraft.server.Main.main(Main.java:156) ~[server-intermediary.jar:?]
>         ... 5 more
> ```

this is a bug with minecraft's 1.19.3+ changes itself, and i'm pretty sure i know why this is happening

**Comment by AnOpenSauceDev** [comment]
apart from 1.19.4+'s breaking changes, this issue is mostly solved with `1.5` and the upcoming `1.6` and it's fallback weights and datapack integration.
